### PR TITLE
Refactored Google Drive Connector + Permission Syncing

### PR DESCRIPTION
**Pull request by hagen-danswer**
## Description
Now, when using either auth flow in google, you will be able to select whether or not you want to index personal drives.

This is especially notable for orgs using the service account auth flow with permission syncing because now everyones docs can be synced and the permissions will match those in the source

### CODE REVIEW

File: web/src/app/admin/connectors/[connector]/pages/gdrive/Credential.tsx [hunk 1]
```diff
@@ -356,23 +356,23 @@ export const DriveOAuthSection = ({
     return (
       <div>
         <p className="text-sm mb-6">
-          When using a Google Drive Service Account, you can either have Danswer
-          act as the service account itself OR you can specify an account for
-          the service account to impersonate.
+          When using a Google Drive Service Account, you must speicify the email
```
Comment (wnmurphy) [inline web/src/app/admin/connectors/[connector]/pages/gdrive/Credential.tsx]: NIT: typo

**Comment by wnmurphy** [comment]
It looks like this adds option to specify URL to a shared drive, is that right?

Excited for Google Drive connector enhancements, thanks!

File: web/src/app/admin/connectors/[connector]/pages/gdrive/Credential.tsx (hunk 1 above)
Comment (hagen-danswer) [reply web/src/app/admin/connectors/[connector]/pages/gdrive/Credential.tsx]: good catch!

File: backend/danswer/connectors/google_drive/connector.py [hunk 2]
```diff
@@ -1,556 +1,255 @@
...
+    def _fetch_drive_items(
         self,
+        is_slim: bool,
         start: SecondsSinceUnixEpoch | None = None,
         end: SecondsSinceUnixEpoch | None = None,
-    ) -> GenerateDocumentsOutput:
-        if self.creds is None:
-            raise PermissionError("Not logged into Google Drive")
+    ) -> Iterator[GoogleDriveFileType]:
+        admin_drive_service = self.get_google_resource()
```
Comment (Weves) [inline backend/danswer/connectors/google_drive/connector.py]: improve naming, this overlaps with `admin_service`

File: backend/danswer/connectors/google_drive/connector.py [hunk 3]
```diff
@@ -1,556 +1,255 @@
...
-            return text
-        elif mime_type == GDriveMimeType.POWERPOINT.value:
-            return pptx_to_text(file=io.BytesIO(response))
+        self.include_my_drives = include_my_drives
+        self.my_drive_emails = _get_string_list_from_comma_separated_string(
+            my_drive_emails
+        )
 
-    return UNSUPPORTED_FILE_TYPE_CONTENT
+        self.folder_ids = _get_string_list_from_comma_separated_string(folder_ids)
```
Comment (Weves) [inline backend/danswer/connectors/google_drive/connector.py]: throw error if this is specified

File: backend/danswer/connectors/google_drive/connector.py [hunk 4]
```diff
@@ -1,556 +1,255 @@
...
+            all_user_emails: set[str] = set(self.my_drive_emails or [])
+
+            # If using service account and no emails specified, fetch all users
+            if not all_user_emails and isinstance(
+                self.creds, ServiceAccountCredentials
+            ):
+                all_user_emails = set(self._get_all_user_emails())
+
+            # Always include the primary admin email
+            if self.primary_admin_email:
```
Comment (Weves) [inline backend/danswer/connectors/google_drive/connector.py]: only include if oauth

File: backend/danswer/connectors/google_drive/constants.py [hunk 5]
```diff
@@ -1,7 +1,29 @@
...
+SLIM_FILE_FIELDS = "nextPageToken, files(id, permissions(emailAddress, type), permissionIds, webViewLink)"
+FOLDER_FIELDS = "nextPageToken, files(id, name, permissions, modifiedTime, webViewLink, shortcutDetails)"
+USER_FIELDS = "nextPageToken, users(primaryEmail)"
+
+# these errors don't represent a failure in the connector, but simply files
+# that can't / shouldn't be indexed
+ERRORS_TO_CONTINUE_ON = [
+    "cannotExportFile",
+    "exportSizeLimitExceeded",
+    "cannotDownloadFile",
```
Comment (wnmurphy) [inline backend/danswer/connectors/google_drive/constants.py:22]: What would you think about adding `notFound`?

Am trying to index a shared drive with a service account impersonating a workspace user, and indexing is hanging on these while they retry every 30s:
```
googleapiclient.errors.HttpError: <HttpError 404 when requesting https://www.googleapis.com/drive/v3/files/1YgCxrbFsFS1ZQBdsialmpe_jVs4MMBLmTJe9sV22uZw?supportsAllDrives=true&alt=json returned "File not found: 1YgCxrbFsFS1ZQBdsialmpe_jVs4MMBLmTJe9sV22uZw.". Details: "[{'message': 'File not found: 1YgCxrbFsFS1ZQBdsialmpe_jVs4MMBLmTJe9sV22uZw.', 'domain': 'global', 'reason': 'notFound', 'location': 'fileId', 'locationType': 'parameter'}]">
```

File: web/src/lib/connectors/connectors.tsx [hunk 6]
```diff
@@ -204,38 +211,58 @@ export const connectorConfigs: Record<
...
+          values.include_my_drives &&
+          !currentCredential?.credential_json?.google_drive_tokens,
+        optional: true,
+      },
+    ],
+    advanced_values: [
+      {
+        type: "text",
+        description:
+          "Enter a comma separated list of the IDs of the folders to index:",
```
Comment (Weves) [inline web/src/lib/connectors/connectors.tsx]: Use urls + make textarea

File: web/src/lib/connectors/connectors.tsx [hunk 7]
```diff
@@ -204,38 +211,58 @@ export const connectorConfigs: Record<
...
       {
-        type: "checkbox",
-        query: "Include shared files?",
-        label: "Include Shared",
-        name: "include_shared",
-        optional: false,
-        default: false,
+        type: "text",
+        description:
+          "Enter a comma separated list of the IDs of the shared drives to index. Leave blank to index all shared drives.",
```
Comment (Weves) [inline web/src/lib/connectors/connectors.tsx]: Make urls + use `textarea` for wrapping

File: web/src/lib/connectors/connectors.tsx [hunk 8]
```diff
@@ -204,38 +211,58 @@ export const connectorConfigs: Record<
     description: "Configure Google Drive connector",
     values: [
       {
-        type: "list",
-        query: "Enter folder paths:",
-        label: "Folder Paths",
-        name: "folder_paths",
+        type: "checkbox",
+        description: "Include shared drives?",
```
Comment (Weves) [inline web/src/lib/connectors/connectors.tsx]: remove redundant description or add something more interesting

**Review by Weves — CHANGES_REQUESTED** [review]

File: backend/danswer/connectors/google_drive/constants.py (hunk 5 above)
Comment (hagen-danswer) [reply backend/danswer/connectors/google_drive/constants.py:22]: Try again once this PR is in. The flow is:
If include_shared_drives is True, the service account impersonates all (or indicated) users and indexes their personal drive.
Then, to access shared drives, it impersonates the primary admin email to access all drives that admin has access to

**Review by Weves — APPROVED** [review]

File: backend/tests/daily/connectors/google_drive/file_generator.py
Comment (Weves) [inline backend/tests/daily/connectors/google_drive/file_generator.py:1]: this should not be here, right?

File: backend/tests/daily/connectors/google_drive/test_google_drive_slim_docs.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            accessible_file_names_for_user.append(file_name)
+        elif any(
+            group_email in external_access.external_user_group_ids
+            for group_email in group_emails_for_user
+        ):
+            accessible_file_names_for_user.append(file_name)
+    return accessible_file_names_for_user
+
+
+def check_access_for_user(
```
Comment (Weves) [inline backend/tests/daily/connectors/google_drive/test_google_drive_slim_docs.py]: for future reference, would prefer if these tests were a bit less abstracted. imo It's pretty hard to tell what's going on here (and throughout the test)

Comment (Weves) [reply backend/tests/daily/connectors/google_drive/test_google_drive_slim_docs.py]: let's chat through this (maybe as a team). but not worth re-writing these ones

**Review by Weves — APPROVED** [review]

**Review by Weves — APPROVED** [review]

**Pull request merged**